go/protomodule/protomodule_list.go

```
@@ -0,0 +1,191 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package protomodule
+
+import (
+	"fmt"
+
+	"go.starlark.net/starlark"
+	"go.starlark.net/syntax"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+var allowedListMethods = map[string]func(*protoRepeated) starlark.Value{
+	"clear":  nil,
+	"append": (*protoRepeated).wrapAppend,
+	"extend": (*protoRepeated).wrapExtend,
+}
+
+// protoRepeated wraps an underlying starlark.List to provide typechecking on
+// wrties
+//
+// starlark.List is heterogeneous, where protoRepeated enforces all values
+// conform to the given fieldDesc
+type protoRepeated struct {
+	fieldDesc protoreflect.FieldDescriptor
+	list      *starlark.List
+}
+
+var _ starlark.Value = (*protoRepeated)(nil)
+var _ starlark.Iterable = (*protoRepeated)(nil)
+var _ starlark.Sequence = (*protoRepeated)(nil)
+var _ starlark.Indexable = (*protoRepeated)(nil)
+var _ starlark.HasAttrs = (*protoRepeated)(nil)
+var _ starlark.HasSetIndex = (*protoRepeated)(nil)
+var _ starlark.HasBinary = (*protoRepeated)(nil)
+var _ starlark.Comparable = (*protoRepeated)(nil)
+
+func newProtoRepeated(fieldDesc protoreflect.FieldDescriptor) *protoRepeated {
+	return &protoRepeated{fieldDesc, starlark.NewList(nil)}
+}
+
+func newProtoRepeatedFromList(fieldDesc protoreflect.FieldDescriptor, l *starlark.List) (*protoRepeated, error) {
+	out := &protoRepeated{fieldDesc, l}
+	for i := 0; i < l.Len(); i++ {
+		err := scalarTypeCheck(fieldDesc, l.Index(i))
+		if err != nil {
+			return nil, err
+		}
+	}
+	return out, nil
+}
+
+func (r *protoRepeated) Attr(name string) (starlark.Value, error) {
+	wrapper, ok := allowedListMethods[name]
+	if !ok {
+		return nil, nil
```

OOC, is there a reason we don't want to return an error here? The intention/context here might be comment-worthy.

---

I had to look this up https://pkg.go.dev/go.starlark.net/starlark?utm_source=godoc#HasAttrs
> For implementation convenience, a result of (nil, nil) from Attr is interpreted as a "no such field or method" error. Implementations are free to return a more precise error.